quiche/src/stream/mod.rs

```diff
@@ -816,44 +816,41 @@ impl PartialEq for StreamPriorityKey {
 impl Eq for StreamPriorityKey {}
 
 impl PartialOrd for StreamPriorityKey {
-    // Priority ordering is complex, disable Clippy warning.
-    #[allow(clippy::non_canonical_partial_ord_impl)]
     fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for StreamPriorityKey {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
         // Ignore priority if ID matches.
         if self.id == other.id {
-            return Some(cmp::Ordering::Equal);
+            return cmp::Ordering::Equal;
         }
 
         // First, order by urgency...
         if self.urgency != other.urgency {
-            return self.urgency.partial_cmp(&other.urgency);
+            return self.urgency.cmp(&other.urgency);
         }
 
         // ...when the urgency is the same, and both are not incremental, order
         // by stream ID...
         if !self.incremental && !other.incremental {
-            return self.id.partial_cmp(&other.id);
+            return self.id.cmp(&other.id);
         }
 
         // ...non-incremental takes priority over incremental...
         if self.incremental && !other.incremental {
-            return Some(cmp::Ordering::Greater);
+            return cmp::Ordering::Greater;
         }
         if !self.incremental && other.incremental {
-            return Some(cmp::Ordering::Less);
+            return cmp::Ordering::Less;
         }
 
         // ...finally, when both are incremental, `other` takes precedence (so
         // `self` is always sorted after other same-urgency incremental
         // entries).
-        Some(cmp::Ordering::Greater)
-    }
-}
-
-impl Ord for StreamPriorityKey {
-    fn cmp(&self, other: &Self) -> cmp::Ordering {
-        // `partial_cmp()` never returns `None`, so this should be safe.
-        self.partial_cmp(other).unwrap()
+        cmp::Ordering::Greater
```

Comment:
I don't know what the requirements for cmp are and what are the consequences of having elements that return different answers if you swap the argument order.

I think this is a case where the id could be compared.

In the partial order case, this case may want to return None.

Reply:
That would break the prioritization algorithm which needs these types of streams to be round-robined. This is a bit of a trick, newly added incremental streams of the same urgency to "go to the back", so by removing then adding a stream when it's been visited (e.g. to send data) we can ensure the next visit hits the next stream of this type.